lib/operations/crud_write.js

```diff
@@ -0,0 +1,284 @@
+'use strict';
+
+const operation = require('./operation');
+const Aspect = operation.Aspect;
+const defineAspects = operation.defineAspects;
+const OperationBase = operation.OperationBase;
+const applyRetryableWrites = require('../utils').applyRetryableWrites;
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const handleCallback = require('../utils').handleCallback;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const toError = require('../utils').toError;
+
+const INSERT = 'insert';
+const UPDATE = 'update';
+const REMOVE = 'remove';
+
+const RETRYABLE_WIRE_VERSION = 6;
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+function isRetryabilitySupported(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+}
+
+class CrudWriteOperation extends OperationBase {
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
+    return (
+      this.options.retryWrites &&
+      this.session &&
+      !this.session.inTransaction() &&
+      isRetryabilitySupported(topology) &&
+      this.model.canRetry()
+    );
+  }
+
+  enableRetry() {
+    this.model.options.willRetryWrite = true;
+    this.session.incrementTransactionNumber();
+  }
+
+  execute(server, callback) {
+    return this.model.execute(server, callback);
+  }
+}
+
+class CrudModel {
+  constructor(collection, options) {
+    options = applyRetryableWrites(options, collection.s.db);
```

Comment:
seems like a pretty prime location to make a shallow copy of the options?

Reply:
I was following @kvwalker's advice [here](https://github.com/mongodb/node-mongodb-native/pull/1991#discussion_r282580377). Currently we already make the shallow copy in the collection, and will later handle it with options management.